src/aws-codeartifact/domain.ts

```diff
@@ -0,0 +1,213 @@
+  readonly domainOwner: string;
+}
+
+export interface DomainProps {
+  /**
+   * The name of the Domain
+   */
+  readonly domainName: string;
+  /**
+   * The key used to encrypt the Domain
+   */
+  readonly encryptionKey?: IKey;
+}
+
+export class Domain extends DomainBase implements IDomain, ITaggableV2 {
+  public static fromDomainAttributes(scope: Construct, id: string, attrs: DomainAttributes): IDomain {
+    class Import extends DomainBase {
+      public readonly domainArn = attrs.domainArn;
+      public readonly domainName = attrs.domainName;
+      public readonly encryptionKeyArn = attrs.encryptionKeyArn;
+      public readonly domainOwner = attrs.domainOwner;
+    }
+
+    return new Import(scope, id);
+  }
+
+  public static fromDomainArn(scope: Construct, id: string, domainArn: string): IDomain {
+    const domainResourceArnParts = Stack.of(scope).splitArn(domainArn, ArnFormat.SLASH_RESOURCE_NAME);
+    if (
+      domainResourceArnParts.resource !== 'domain' ||
+      domainResourceArnParts.account === undefined ||
+      domainResourceArnParts.resourceName === undefined
+    ) {
+      throw new Error(`Expected a domain ARN, but got ${domainArn}`);
+    }
+    return Domain.fromDomainAttributes(scope, id, {
+      domainArn,
+      domainName: domainResourceArnParts.resourceName,
+      domainOwner: domainResourceArnParts.account,
+    });
+  }
+
+  protected cfnResource: CfnDomain;
+  private cfnResourceProps: CfnDomainProps;
+
+  readonly cdkTagManager: TagManager;
+  readonly domainArn: string;
+  readonly domainName: string;
+  readonly encryptionKeyArn?: string;
+  readonly domainOwner: string;
+  /**
+   * Optional policy document that represents the resource policy of this repository
+   *
+   * If specified, addToResourcePolicy can be used to edit this policy.
+   * Otherwise this method will no-op.
+   */
+  protected readonly policy?: PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: DomainProps) {
+    super(scope, id);
+    this.policy = new PolicyDocument();
+    this.cdkTagManager = new TagManager(TagType.KEY_VALUE, 'AWS::CodeArtifact::Domain');
+
+    const encryptionKey =
+      props.encryptionKey ??
+      new Key(this, 'key', {
+        description: `Key for CodeArtifact Domain ${props.domainName}`,
+        alias: `codeartifact-domain/${props.domainName}`,
+      });
+
+    this.cfnResourceProps = {
+      domainName: props.domainName,
+      encryptionKey: encryptionKey?.keyArn,
+      tags: this.cdkTagManager.renderTags(),
+      permissionsPolicyDocument: this.policy,
+    };
+    this.cfnResource = this.createCfnResource();
+
+    this.domainName = this.cfnResource.attrName;
+    this.domainArn = this.cfnResource.attrArn;
+    this.encryptionKeyArn = this.cfnResource.attrEncryptionKey;
+    this.domainOwner = this.cfnResource.attrOwner;
+  }
+
+  protected createCfnResource(): CfnDomain {
+    return new CfnDomain(this, 'Resource', this.cfnResourceProps);
```

Comment:
```suggestion
    return new CfnDomain(this, 'Default', this.cfnResourceProps);
```

Reply:
Aren't we using Resource for the defaultChild thoughout this library and CDK Core?

Reply:
I guess it doesn't make a difference. The way `Node.defaultChild()` is [implemented](https://github.com/aws/constructs/blob/f0b8dc8025efa0b0c30bd3a9894d60f1b29c04da/src/construct.ts#L138-L158), both work.
Not sure this will change as some time. But in the aws-cdk usually `Resource` will be used for the defaultChild and since it is a resource we're creating here (and this might at some point get integrated upstream) I'd also stick to that.